### PR TITLE
[fix] Add blob typechecker

### PR DESCRIPTION
**Pull request by ddl-rliu**
## Tracking issue
https://github.com/flyteorg/flyte/issues/5553

Likely similar issue to https://github.com/flyteorg/flyte/issues/2864

Repro idea: `pyflyte run --remote wf.py wf`
```py
import csv
from collections import defaultdict
from pathlib import Path
from typing import List, TypeVar, NamedTuple

import flytekit
from flytekit import task, workflow
from flytekit.types.file import FlyteFile


@task(cache=True, cache_version="3.0")
def t1() -> FlyteFile[TypeVar("csv")]:
    columns_to_normalize = ["a", "b"]
    out_path = Path(flytekit.current_context().working_directory) / f"normalized.csv"
    with out_path.open(mode="w") as output_file:
        writer = csv.DictWriter(output_file, fieldnames=columns_to_normalize)
        writer.writeheader()

    print(columns_to_normalize)
    return FlyteFile(path=out_path)


@task(cache=True, cache_version="3.0")
def t2(ff: FlyteFile[TypeVar("csv")]) -> FlyteFile[TypeVar("csv")]:
    print("hello")
    return ff


@workflow
def wf() -> FlyteFile[TypeVar("csv")]:
    return t2(ff=t1())


if __name__ == "__main__":
    wf()
```

## Why are the changes needed?

An error happens whenever a `FlyteFile[TypeVar("csv")]` output is passed to another input, with caching enabled.

Errors in flytepropeller when caching is on for FlyteFile type:
`{"json":{"exec_id":"legally-working-squirrel-zel6-zrvvdwz7jrrf79cwm44j","node":"n0","ns":"domino-compute","res_ver":"73517980","routine":"worker-1","wf":"6668e7c9d14c4306e8cc54d1:development:workflow.training_workflow_cache"},"level":"error","msg":"DataCatalog failed to get outputs from artifact adbf03db-daea-4108-9f8f-c352ac395a6b, err: unexpected artifactData: [processed_data] type: [blob:{}] does not match any task output type: [blob:{format:\"csv\"}]","ts":"2024-06-26T00:05:39Z"}`

## What changes were proposed in this pull request?

I believe these errors happen when the `trivialChecker` is used for blob types. This PR adds a `blobTypeChecker`.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

File: flytepropeller/pkg/compiler/validators/typing_test.go [hunk 1]
```diff
@@ -893,3 +893,10 @@ func TestStructuredDatasetCasting(t *testing.T) {
 		assert.True(t, castable, "StructuredDataset are nullable")
 	})
 }
+
+// TODO add tests for blob typechecker
+//
+// WIP test suite
+// assert.True(t, castable, "Blob(format='csv') should be castable to Blob()")
+// assert.True(t, castable, "Blob() should be castable to Blob()")
+// assert.False(t, castable, "Blob(format='csv') should not be castable to Blob('zip')")
```
Comment (ddl-rliu) [inline flytepropeller/pkg/compiler/validators/typing_test.go:902]: note-to-self: understand this test file and understand how typechecker works in order to define the correct spec for the new blob typechecker

File: flytepropeller/pkg/compiler/validators/typing_test.go [hunk 2]
```diff
@@ -893,3 +893,58 @@ func TestStructuredDatasetCasting(t *testing.T) {
...
+	t.Run("BlobsAreNullable", func(t *testing.T) {
+		castable := AreTypesCastable(
+			&core.LiteralType{
+				Type: &core.LiteralType_Simple{
+					Simple: core.SimpleType_NONE,
+				},
+			},
+			genericBlob)
+		assert.False(t, castable, "Blob is not nullable")
+	})
```
Comment (ddl-rliu) [inline flytepropeller/pkg/compiler/validators/typing_test.go:949]: These test cases are my best approximation of what behavior a Blob typechecker should have, based on the other test cases in this suite.

File: flytepropeller/pkg/compiler/validators/typing.go [hunk 3]
```diff
@@ -72,6 +72,26 @@ func (t mapTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
...
 }
 
+type blobTypeChecker struct {
+	literalType *flyte.LiteralType
+}
+
+// CastsFrom checks that the target blob type can be cast to the current blob type. When the blob has no format
+// specified, it accepts all blob inputs since it is generic.
+func (t blobTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
+	blobType := upstreamType.GetBlob()
```
[the remark this answers is not in the document]
Comment (hamersaw) [reply flytepropeller/pkg/compiler/validators/typing.go:82]: Great point, it doesn't look like any other type checkers validate nil inputs. In a brief dive, IIUC it is impossible for upstream to be `nil` here. I do not have concerns leaving out the nil check.

**Review by hamersaw — COMMENTED** [review]
This looks reasonable to me. Checking with flytekit folks if there are any concerns here from DevX perspective (cc @pingsutw @eapolinario ). Then happy to merge!

**Review by pingsutw — APPROVED** [review]
LGTM, thank you!

**Pull request merged**